examples/forward_backward.py

```diff
@@ -0,0 +1,187 @@
+    """
+    step = OrderedDict(sorted(step.items()))
+    drop = tuple("_drop_{}".format(i) for i in range(len(step)))
+    prev_to_drop = dict(zip(step.keys(), drop))
+    curr_to_drop = dict(zip(step.values(), drop))
+    reduce_vars = frozenset(drop)
+
+    factors = [trans(**{time: t}) for t in range(trans.inputs[time].size)]
+    # base case
+    inputs = OrderedDict((k, v) for k, v in trans.inputs.items() if k in step.keys())
+    data = funsor.ops.new_zeros(funsor.tensor.get_default_prototype(), ()).expand(
+        tuple(v.size for v in inputs.values())
+    )
+    data = data + UNITS[prod_op]
+    base = Tensor(data, inputs, trans.dtype)
+    betas = [base]
+    # inductive steps
+    while factors:
+        x = factors.pop()(**curr_to_drop)
+        base = prod_op(x, base(**prev_to_drop)).reduce(sum_op, reduce_vars)
+        betas.append(base)
+    else:
+        base = prod_op(init(**curr_to_drop), base(**prev_to_drop)).reduce(
+            sum_op, reduce_vars
+        )
+    betas.reverse()
+    return base, betas
+
+
+def main(args):
+    """
+    Compute smoothed values (gamma[t] = p(x[t]|y[0:T])) for an HMM:
+
+    x[0] -> ... -> x[t-1] -> x[t] -> ... -> x[T]
+     |              |         |             |
+     v              v         v             v
+    y[0]           y[t-1]    y[t]           y[T]
+
+    **References:**
+
+    [1] Jason Eisner (2016)
+        "Inside-Outside and Forward-Backward Algorithms Are Just Backprop
+        (Tutorial Paper)"
+        https://www.cs.jhu.edu/~jason/papers/eisner.spnlp16.pdf
+    [2] Zhifei Li and Jason Eisner (2009)
+        "First- and Second-Order Expectation Semirings
+        with Applications to Minimum-Risk Training on Translation Forests"
+        http://www.cs.jhu.edu/~zfli/pubs/semiring_translation_zhifei_emnlp09.pdf
+    """
+    funsor.set_backend("torch")
+    sum_op = ops.logaddexp
+    prod_op = ops.add
+
+    # transition and emission probabilities
+    time = "time"
+    init = random_tensor(OrderedDict([("x_curr", Bint[args.hidden_dim])]))
+    trans = random_tensor(
+        OrderedDict(
+            [
+                (time, Bint[args.time_steps]),
+                ("x_prev", Bint[args.hidden_dim]),
+                ("x_curr", Bint[args.hidden_dim]),
+            ]
+        )
+    )
+
+    # Compute smoothed values by using the forward-backward algorithm
+    # gamma[t] = p(x[t]|y[0:T])
+    # gamma[t] = alpha[t] * beta[t]
+    Z_forward, alphas = forward_algorithm(
+        sum_op, prod_op, init, trans, time, {"x_prev": "x_curr"}
+    )
+    Z_backward, betas = backward_algorithm(
+        sum_op, prod_op, init, trans, time, {"x_prev": "x_curr"}
+    )
+    assert_close(Z_forward, Z_backward)
+    gammas = []
+    for alpha, beta in zip(alphas, betas):
+        gamma = prod_op(alpha(**{"x_curr": "x_prev"}), beta)
+        gammas.append(gamma)
+
+    # forward-backward algorithm can be obtained by differentiating the backward algorithm
```

Comment:
I suppose this is true, but it seems like the point we'd like to make with this example is the same one made in reference [1], which is that the marginal probabilities `p(x[t] | Y)` or `p(x[t], x[t-1] | Y)` can be obtained directly by differentiating `Z_forward`, without ever constructing `beta[t]` ourselves - these are the quantities we actually care about, after all.

It would be easier to obtain these if you broke `trans` up into separate (log-)factors `p(x[t] | x[t-1])`, `p(y[t] | x[t])` in `main`.  Then you could write something like:
```py
with AdjointTape() as tape:
    log_Z = forward_algorithm(sum_op, prod_op, [trans, obs], time, {"x_prev": "x_curr"})  # don't return alphas

adjoint_log_marginals = tape.adjoint(sum_op, prod_op, log_Z, [trans, obs])
```
You could make `backward_algorithm` into `forward_backward_algorithm` that computes the same things by hand and compare the resulting marginals.

Reply:
I see. It makes more sense to me now. With the changes it will be much better example. Thanks!